File: stepik-union/src/main/java/org/stepik/plugin/actions/ActionUtils.java
package org.stepik.plugin.actions;

import com.intellij.execution.ui.ConsoleViewContentType;
import com.intellij.openapi.actionSystem.KeyboardShortcut;
import com.intellij.openapi.keymap.KeymapUtil;
import org.jetbrains.annotations.NotNull;
import org.stepik.api.objects.submissions.Submission;
import org.stepik.core.testFramework.toolWindow.StepikTestResultToolWindow;

import javax.swing.*;

public class ActionUtils {
    static final int MILLISECONDS_IN_MINUTES = 60 * 1000;
    private static final int MILLISECONDS_IN_HOUR = 60 * MILLISECONDS_IN_MINUTES;
    private static final String HOUR = "hour";
    private static final String MINUTE = "minute";
    private static final String SECOND = "second";

    public static String etaAsString(long eta) {
        StringBuilder result = new StringBuilder();

        long hours = eta / MILLISECONDS_IN_HOUR;
        if (hours > 0) {
            appendTimePart(result, hours, HOUR);
            eta = eta - hours * MILLISECONDS_IN_HOUR;
        }

        long minutes = eta / MILLISECONDS_IN_MINUTES;
        if (minutes > 0) {
            appendTimePart(result, minutes, MINUTE);
            eta = eta - minutes * MILLISECONDS_IN_MINUTES;
        }

        long seconds = eta / 1000;
        if (seconds > 0) {
            appendTimePart(result, seconds, SECOND);
        }

        return result.toString().trim();
    }

    private static void appendTimePart(StringBuilder result, long value, String label) {
        result.append(value)
                .append(" ")
                .append(label)
                .append(value == 1 ? "" : "s")
                .append(" ");
    }

    static void setupCheckProgress(
            @NotNull StepikTestResultToolWindow resultWindow,
            @NotNull Submission submission,
            long timer) {
        double eta = submission.getEta() * 1000; //to ms
        double total = eta + timer;

        resultWindow.print("[" + Math.round((1 - eta / total) * 100) + "%] ", ConsoleViewContentType.NORMAL_OUTPUT);
        resultWindow.print("Ends in " + etaAsString((long) eta), ConsoleViewContentType.NORMAL_OUTPUT);
    }

    @NotNull
    public static String getShortcutText(@NotNull String shortcut) {
        KeyboardShortcut keyboardShortcut = new KeyboardShortcut(KeyStroke.getKeyStroke(shortcut), null);
        return KeymapUtil.getShortcutText(keyboardShortcut);
    }
}

File: stepik-union/src/main/kotlin/org/stepik/core/actions/ActionUtils.kt
package org.stepik.core.actions

import com.intellij.openapi.actionSystem.KeyboardShortcut
import com.intellij.openapi.keymap.KeymapUtil
import org.stepik.api.objects.submissions.Submission
import org.stepik.core.testFramework.toolWindow.StepikTestResultToolWindow
import javax.swing.KeyStroke

const val MILLISECONDS_IN_MINUTES = 60 * 1000
private const val MILLISECONDS_IN_HOUR = 60 * MILLISECONDS_IN_MINUTES
private const val HOUR = "hour"
private const val MINUTE = "minute"
private const val SECOND = "second"

fun etaAsString(eta: Long): String {
    var value = eta
    val result = StringBuilder()

    val hours = value / MILLISECONDS_IN_HOUR
    if (hours > 0) {
        appendTimePart(result, hours, HOUR)
        value -= hours * MILLISECONDS_IN_HOUR
    }

    val minutes = value / MILLISECONDS_IN_MINUTES
    if (minutes > 0) {
        appendTimePart(result, minutes, MINUTE)
        value -= minutes * MILLISECONDS_IN_MINUTES
    }

    val seconds = value / 1000
    if (seconds > 0) {
        appendTimePart(result, seconds, SECOND)
    }

    return result.toString().trim { it <= ' ' }
}

private fun appendTimePart(result: StringBuilder, value: Long, label: String) {
    result.append(value)
            .append(" ")
            .append(label)
            .append(if (value == 1L) "" else "s")
            .append(" ")
}

fun setupCheckProgress(
        resultWindow: StepikTestResultToolWindow,
        submission: Submission,
        timer: Long) {
    val eta = submission.eta * 1000 //to ms
    val total = eta + timer

    resultWindow.print("[${Math.round((1 - eta / total) * 100)}%] ")
    resultWindow.print("Ends in ${etaAsString(eta.toLong())}")
}

fun getShortcutText(shortcut: String): String {
    val keyboardShortcut = KeyboardShortcut(KeyStroke.getKeyStroke(shortcut), null)
    return KeymapUtil.getShortcutText(keyboardShortcut)
}
